internal/backend/runtime/omni/controllers/omni/reporting.go

```diff
@@ -0,0 +1,158 @@
+	"github.com/cosi-project/runtime/pkg/state"
+	"github.com/stripe/stripe-go/v74"
+	"github.com/stripe/stripe-go/v74/subscriptionitem"
+	"go.uber.org/zap"
+
+	"github.com/siderolabs/omni/client/pkg/omni/resources"
+	"github.com/siderolabs/omni/client/pkg/omni/resources/omni"
+)
+
+// DebounceDuration defines how long to wait after the last event before processing updates.
+var DebounceDuration = 10 * time.Minute
+
+// StripeMetricsReporterControllerName is the name of the StripeMetricsReporterController.
+const StripeMetricsReporterControllerName = "StripeMetricsReporterController"
+
+// StripeMetricsReporterController ...
+type StripeMetricsReporterController struct {
+	generic.NamedController
+
+	stripeAPIKey             string
+	stripeSubscriptionItemID string
+}
+
+// NewStripeMetricsReporterController initializes StripeMetricsReporterController.
+func NewStripeMetricsReporterController(stripeAPIKey, stripeSubscriptionItemID string) *StripeMetricsReporterController {
+	return &StripeMetricsReporterController{
+		NamedController: generic.NamedController{
+			ControllerName: StripeMetricsReporterControllerName,
+		},
+		stripeAPIKey:             stripeAPIKey,
+		stripeSubscriptionItemID: stripeSubscriptionItemID,
+	}
+}
+
+// Inputs implements controller.QController interface.
+func (ctrl *StripeMetricsReporterController) Inputs() []controller.Input {
+	return []controller.Input{
+		{
+			Namespace: resources.DefaultNamespace,
+			Type:      omni.MachineStatusMetricsType,
+			Kind:      controller.InputWeak,
+		},
+	}
+}
+
+// Outputs implements controller.Controller interface.
+func (ctrl *StripeMetricsReporterController) Outputs() []controller.Output {
+	return nil
+}
+
+// Run implements controller.Controller interface.
+func (ctrl *StripeMetricsReporterController) Run(ctx context.Context, r controller.Runtime, _ *zap.Logger) error {
+	stripe.Key = ctrl.stripeAPIKey
+
+	var (
+		pendingCount uint32
+		timer        *time.Timer
+		timerCh      <-chan time.Time
+	)
+
+	processPending := func() {
+		count := pendingCount
+		pendingCount = 0
+
+		err := updateStripeSubscriptionItemQuantity(ctrl.stripeSubscriptionItemID, count)
+		if err != nil {
+			log.Printf("Failed to update subscription item %s with count %d: %v", ctrl.stripeSubscriptionItemID, count, err)
+		}
+	}
+
+	for {
+		select {
+		case <-ctx.Done():
+			if timer != nil {
+				timer.Stop()
+			}
+
+			processPending()
+
+			return nil
+		case <-r.EventCh():
+		case <-timerCh:
+			processPending()
+
+			timer.Stop()
+
+			timerCh = nil
+		}
+
+		metrics, err := safe.ReaderGetByID[*omni.MachineStatusMetrics](ctx, r, omni.MachineStatusMetricsID)
+		if err != nil {
+			if state.IsNotFoundError(err) {
+				continue
+			}
+
+			return err
+		}
+
+		pendingCount = metrics.TypedSpec().Value.RegisteredMachinesCount
+
+		if timer != nil {
+			if !timer.Stop() {
+				<-timer.C
+			}
+		}
+
+		timer = time.NewTimer(DebounceDuration)
+
+		timerCh = timer.C
+	}
```

Comment:
As a common practice, I suggest putting `r.ResetRestartBackoff()` as the last statement in the `for` loop of events.